# 🧮 SQL-Driven Exploration: CMS Hospital Dataset

This notebook showcases my SQL expertise through a structured, self-directed analysis of the **CMS Hospital General Information** dataset — a public dataset from the Centers for Medicare & Medicaid Services (CMS):  
🔗 [CMS Hospital General Information Dataset](https://data.cms.gov/provider-data/dataset/xubh-q36u)

The dataset includes detailed information about over 5,000 hospitals in the United States, such as hospital types, ownership structures, performance metrics, and quality ratings.

---

### 🎯 Project Goal

The primary goal of this notebook is to demonstrate practical SQL proficiency by performing meaningful, interview-relevant analysis **entirely through SQL**. All queries were written from scratch, based on questions I designed myself to explore patterns, trends, and relationships in the data.

---

### 🧠 What This Notebook Shows

- My ability to **write clean, modular, and well-commented SQL code**
- Critical thinking through **self-directed question framing**
- Comfort with a range of SQL operations:  
  ✅ Filtering and aggregation  
  ✅ Window functions and ranking  
  ✅ Set operations (e.g., `UNION`, `EXCEPT`)  
  ✅ CASE logic and derived metrics  
  ✅ String and date functions  
  ✅ Joins and subqueries  
  ✅ Pivot-style queries with conditional aggregation  

Each section includes:
- A clear question framed in business or stakeholder language  
- A SQL query executed directly on a PostgreSQL database  
- A short summary interpreting the output in plain language  

---

> This notebook represents a real-world analyst workflow — not just SQL syntax practice — and is intended to show hiring managers both **technical depth** and **analytical thought process**.

---
## 📑 Table of Contents

- [🧮 Query 1: Hospital Count by State](#🧮-Query-1:-Hospital-Count-by-State)
- [🧮 Query 2: Hospital Count by Type](#🧮-Query-2:-Hospital-Count-by-Type)
- [🧮 Query 3: Emergency Services by State](#🧮-Query-3:-Emergency-Services-by-State)
- [🧮 Query 4: Hospitals Missing Overall Rating by State](#🧮-Query-4:-Hospitals-Missing-Overall-Rating-by-State)
- [🧮 Query 5: Emergency Services Availability by Hospital Type](#🧮-Query-5:-Emergency-Services-Availability-by-Hospital-Type)
- [🧮 Query 6: Average Rating by Ownership Type](#🧮-Query-6:-Average-Rating-by-Ownership-Type)
- [🧮 Query 7: Hospitals with 5-Star Rating by State](#🧮-Query-7:-Hospitals-with-5-Star-Rating-by-State)
- [🧮 Query 8: Search Hospitals by Name (Keyword: "Children")](#🧮-Query-8:-Search-Hospitals-by-Name-(Keyword:-"Children"))
- [🧮 Query 9: Average Hospital Rating by City (with ≥ 5 Hospitals)](#🧮-Query-9:-Average-Hospital-Rating-by-City-(with-≥-5-Hospitals))
- [🧮 Query 10: Facility Names Shared Across States](#🧮-Query-10:-Facility-Names-Shared-Across-States)
- [🧮 Query 11: Duplicate Facility Names Across States (via Self Join)](#🧮-Query-11:-Duplicate-Facility-Names-Across-States-(via-Self-Join))
- [🧮 Query 12: Duplicate Facility Names Within the Same State](#🧮-Query-12:-Duplicate-Facility-Names-Within-the-Same-State)
- [🧮 Query 13: States with Above-National-Average Hospital Ratings](#🧮-Query-13:-States-with-Above-National-Average-Hospital-Ratings)
- [🧮 Query 14: Top-Rated Hospitals in Each State Using `RANK()`](#🧮-Query-14:-Top-Rated-Hospitals-in-Each-State)
- [🧮 Query 15: Hospitals in the 10 Most Common Cities](#🧮-Query-15:-Hospitals-in-the-10-Most-Common-Cities)
- [🧮 Query 16: Hospitals Grouped into 4 Performance Tiers](#🧮-Query-16:-Hospitals-Grouped-into-4-Performance-Tiers)
- [🧮 Query 17: States with the Most 1-Star Hospitals Using a CTE](#🧮-Query-17:-States-with-the-Most-1-Star-Hospitals-Using-a-CTE)
- [🧮 Query 18: Top-Rated Hospitals Within Each City](#🧮-Query-18:-Top-Rated-Hospitals-Within-Each-City)
- [🧮 Query 19: Most Recent Facility in Each City](#🧮-Query-19:-Most-Recent-Facility-in-Each-City)
- [🧮 Query 20: Cities with the Widest Range in Hospital Ratings](#🧮-Query-20:-Cities-with-the-Widest-Range-in-Hospital-Ratings)
- [🧮 Query 21: Simulated Filtering of Facilities Joined After 2015](#🧮-Query-21:-Simulated-Filtering-of-Facilities-Joined-After-2015)
- [🧮 Query 22: Analyze Name Patterns Using SQL String Functions](#🧮-Query-22:-Analyze-Name-Patterns-Using-SQL-String-Functions)
- [🧮 Query 23: Compare 5-Star Hospitals Offering Emergency Services Using EXCEPT and UNION](#🧮-Query-23:-Compare-5-Star-Hospitals-Offering-Emergency-Services-Using-EXCEPT-and-UNION)
- [🧮 Query 24: Pivot Hospital Ratings into Columns](#🧮-Query-24:-Pivot-Hospital-Ratings-into-Columns)
- [🧮 Query 25: Top States by Average Hospital Rating (Cities with ≥ 3 Hospitals)](#🧮-Query-25:-Top-States-by-Average-Hospital-Rating-(Cities-with-≥-3-Hospitals))
- [🧮 Query 26: Compare Hospitals Across States Using UNION and EXCEPT](#🧮-Query-26:-Compare-Hospitals-Across-States-Using-UNION-and-EXCEPT)
- [🧮 Query 27: Compare Hospital Ratings Within States Using LAG()](#🧮-Query-27:-Compare-Hospital-Ratings-Within-States-Using-LAG())
- [🧮 Query 28: Hospital Ratings Distribution by Ownership Type](#🧮-Query-28:-Hospital-Ratings-Distribution-by-Ownership-Type)

---
## 🧰 Import Libraries

We begin by importing the necessary libraries:

- `pandas` and `numpy` for data handling and inspection
- `sqlalchemy` to establish a connection between Jupyter and the PostgreSQL database

In [1]:
# Import libraries
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
from IPython.display import display, HTML

## 🔌 Connect to PostgreSQL

Before running queries, I set up the PostgreSQL environment using **pgAdmin** and a custom SQL schema file:

- Created a new database named `hospital_er_data`
- Used the script `SQL/create_hospital_table.sql` to define the `hospital_info` table schema
- Imported the cleaned CSV file (`hospital_info_clean.csv`) into the table using pgAdmin’s import tool

This section defines the connection parameters and creates a SQLAlchemy engine to enable querying the local PostgreSQL database directly from this notebook.

In [2]:
# Define database connection settings
db_user = "postgres"
db_pass = "123456"  # My local password
db_host = "localhost"
db_port = "5432"
db_name = "hospital_er_data"

# Create SQLAlchemy engine
engine = create_engine(f"postgresql://{db_user}:{db_pass}@{db_host}:{db_port}/{db_name}")

### 🧮 Query 0: Verify Table Access and Connection

Before starting our main analysis, we run a simple query to confirm that the database connection is working correctly and the `hospital_info` table exists and is accessible.

In [3]:
# Check if connection and table access are working
query_0 = """
SELECT
    COUNT(*) AS total_records
FROM
    hospital_info;
"""

# Execute the query
df_q0 = pd.read_sql_query(query_0, con=engine)

# Display a clean summary
total_records = df_q0.loc[0, "total_records"]
print(f"Total Records in hospital_info Table: {total_records:,}")

Total Records in hospital_info Table: 5,384


---

### 📋 Database Connection Verified

The connection to the PostgreSQL database is active, and the `hospital_info` table has been successfully queried.

**Record Count:** `5,384` rows found in the table.

This confirms that the data was properly imported and is ready for SQL analysis.

---

### 🧮 Query 1: Hospital Count by State

**Business Question**  
How many hospitals are there in each U.S. state?

**Why This Matters**  
Understanding the number of hospitals per state gives a high-level view of geographic distribution across the U.S. This acts as a baseline for identifying regions with high or low hospital coverage, which can later inform deeper analysis on healthcare access, funding, or performance comparisons.

**SQL Concept Introduced**  
- Aggregation with `COUNT(*)`  
- Grouping with `GROUP BY`  
- Sorting results with `ORDER BY`

In [4]:
# Query 1: Count hospitals by state
query_1 = """
SELECT
    "State",
    COUNT(*) AS hospital_count
FROM
    hospital_info
GROUP BY
    "State"
ORDER BY
    hospital_count DESC;
"""

df_q1 = pd.read_sql_query(query_1, con=engine)
df_q1.head(10)  # Preview top 10 states by hospital count

State  hospital_count
0    TX             459
1    CA             379
2    FL             219
3    IL             195
4    OH             194
5    NY             190
6    PA             188
7    LA             156
8    IN             149
9    GA             146

### 📊 Query 1 Summary  
**Texas**, **California**, and **Florida** have the highest number of hospitals, with **Texas (459)** leading by a significant margin. These states likely reflect larger populations and more extensive healthcare networks. This geographic distribution provides a valuable baseline for identifying regional trends and resource allocation in later analyses.

---
### 🧮 Query 2: Hospital Count by Type

**Business Question**  
How many hospitals are there by type (e.g., Acute Care, Psychiatric, Children's)?

**Why This Matters**  
Understanding the distribution of hospital types provides insight into the specialization and services available across facilities. For example, a higher count of acute care hospitals may indicate broader general service coverage, while a lower number of psychiatric or children’s hospitals may reveal service gaps in certain regions.

**SQL Concepts Introduced**  
- Continued use of `GROUP BY`, `COUNT`, and `ORDER BY`

In [5]:
# Query 2: Count of hospitals by hospital type
query_2 = """
SELECT
    "Hospital Type",
    COUNT(*) AS hospital_count
FROM
    hospital_info
GROUP BY
    "Hospital Type"
ORDER BY
    hospital_count DESC;
"""

df_q2 = pd.read_sql_query(query_2, con=engine)
df_q2

Hospital Type  hospital_count
0                  Acute Care Hospitals            3131
1             Critical Access Hospitals            1368
2                           Psychiatric             627
3  Acute Care - Veterans Administration             132
4                             Childrens              94
5    Acute Care - Department of Defense              32

### 📊 Query 2 Summary  
**Acute Care Hospitals (3,131)** make up the majority of hospitals in the dataset, followed by **Critical Access Hospitals (1,368)** and **Psychiatric facilities (627)**.  
Specialized types like **Veterans Administration (132)**, **Children’s (94)**, and **Department of Defense (32)** hospitals are far fewer in number.  
This distribution reflects a national emphasis on general medical care, while specialized services remain limited.

---
### 🧮 Query 3: Emergency Services by State

**Business Question**  
Which U.S. states have the highest number of hospitals that offer emergency services?

**Why this matters**  
Emergency service availability is a critical component of public health infrastructure. Identifying where such services are concentrated helps assess readiness for urgent care and disaster response.

**SQL Concept Introduced**  
- Filtering rows using the `WHERE` clause
- Continued use of `GROUP BY`, `COUNT`, and `ORDER BY`

In [6]:
# Query 3: Emergency Services availability by state
query_3 = """
SELECT
    "State",
    COUNT(*) AS hospital_count
FROM
    hospital_info
WHERE
    "Emergency Services" = 'Yes'
GROUP BY
    "State"
ORDER BY
    hospital_count DESC;
"""

df_q3 = pd.read_sql_query(query_3, con=engine)
df_q3.head(10)  # Preview top 10 rows

State  hospital_count
0    TX             383
1    CA             289
2    FL             183
3    IL             175
4    NY             156
5    PA             148
6    OH             137
7    WI             126
8    IN             118
9    MI             118

### 📊 Query 3 Summary  
States with the highest number of hospitals offering **emergency services** include **Texas (383)**, **California (289)**, and **Florida (183)**. These states also led in total hospital count (Query 1), suggesting that most hospitals in these regions are equipped to handle emergency care — an important consideration for state-level healthcare readiness and planning.

---
### 🧮 Query 4: Hospitals Missing Overall Rating by State

**Business Question**  
Which states have the most hospitals with missing overall quality ratings?

**Why this matters**  
Missing ratings may signal inconsistencies in reporting, data collection gaps, or exclusions from CMS evaluation. Understanding where these gaps are concentrated can inform efforts to improve data completeness and transparency.

**SQL Concepts Introduced**  
- `WHERE IS NULL` filtering to detect missing values
- Continued use of `GROUP BY`, `COUNT`, and `ORDER BY`

In [7]:
# Query 4: Count of hospitals with missing overall ratings by state
query_4 = """
SELECT
    "State",
    COUNT(*) AS hospitals_missing_rating
FROM
    hospital_info
WHERE
    "Hospital overall rating" IS NULL
GROUP BY
    "State"
ORDER BY
    hospitals_missing_rating DESC
"""

df_q4 = pd.read_sql_query(query_4, con=engine)
df_q4.head(10)  # Preview top 10 rows

State  hospitals_missing_rating
0    TX                       264
1    CA                       116
2    LA                       108
3    MN                        97
4    KS                        97
5    IL                        80
6    IA                        80
7    OH                        80
8    OK                        80
9    WI                        76

### 📊 Query 4 Summary  
**Texas** has the highest number of hospitals missing overall quality ratings (**264 facilities**), followed by **California (116)** and **Louisiana (108)**. These gaps in data reporting could signal challenges in transparency, reporting compliance, or data collection infrastructure across states — particularly in regions with a high volume of hospitals.

---
### 🧮 Query 5: Emergency Services Availability by Hospital Type

**Business Question**  
What percentage of hospitals within each hospital type offer emergency services?

**Why this matters**  
Understanding how emergency care availability varies by hospital type reveals potential gaps in urgent care coverage, particularly among specialized or rural facilities. This insight can help assess systemic accessibility across the U.S. healthcare system.

**SQL Concepts Introduced**  
- `CASE WHEN` logic for conditional aggregation  
- `SUM` and `COUNT` to compute service ratios  
- `ROUND` to format percentages

In [8]:
# Query 5: Emergency services availability by hospital type
query_5 = """
SELECT
    "Hospital Type",
    COUNT(*) AS total_hospitals,
    SUM(CASE WHEN "Emergency Services" = 'Yes' THEN 1 ELSE 0 END) AS with_emergency_services,
    ROUND(
        100.0 * SUM(CASE WHEN "Emergency Services" = 'Yes' THEN 1 ELSE 0 END) / COUNT(*),
        1
    ) AS percent_with_emergency
FROM
    hospital_info
GROUP BY
    "Hospital Type"
ORDER BY
    percent_with_emergency DESC;
"""

df_q5 = pd.read_sql_query(query_5, con=engine)
df_q5

Hospital Type  total_hospitals  \
0             Critical Access Hospitals             1368   
1    Acute Care - Department of Defense               32   
2                  Acute Care Hospitals             3131   
3  Acute Care - Veterans Administration              132   
4                             Childrens               94   
5                           Psychiatric              627   

   with_emergency_services  percent_with_emergency  
0                     1330                    97.2  
1                       31                    96.9  
2                     2863                    91.4  
3                      112                    84.8  
4                       55                    58.5  
5                       84                    13.4

### 📊 Query 5 Summary  
**Critical Access Hospitals** and **Acute Care** facilities have the highest availability of emergency services, with **over 90%** of hospitals in each category offering them. In contrast, only **13.4% of Psychiatric hospitals** and **58.5% of Children's hospitals** provide emergency care. These differences highlight how access to urgent medical treatment varies significantly by hospital type, which is critical for healthcare planning and policy decisions.

---
### 🧮 Query 6: Average Rating by Ownership Type

**Business Question**  
How do average hospital ratings differ based on the type of hospital ownership?

**Why this matters**  
Ownership type can influence hospital priorities, patient care, and funding structures. Comparing average CMS ratings by ownership type reveals patterns in performance that may inform policy or investment strategies.

**SQL Concepts Introduced**  
- `AVG()` and `ROUND()` for calculating and formatting averages  
- `WHERE IS NOT NULL` for excluding missing values  
- `GROUP BY` and `ORDER BY` for aggregation and ranking

In [9]:
# Query 6: Average hospital rating by ownership type (excluding missing ratings)
query_6 = """
SELECT
    "Hospital Ownership",
    ROUND(AVG("Hospital overall rating")::numeric, 2) AS avg_rating,
    COUNT(*) AS hospital_count
FROM
    hospital_info
WHERE
    "Hospital overall rating" IS NOT NULL
GROUP BY
    "Hospital Ownership"
ORDER BY
    avg_rating DESC;
"""

df_q6 = pd.read_sql_query(query_6, con=engine)
df_q6  # Display average rating and count by ownership type

Hospital Ownership  avg_rating  hospital_count
0                Veterans Health Administration        3.63             113
1                 Voluntary non-profit - Church        3.34             211
2                  Voluntary non-profit - Other        3.33             241
3                Voluntary non-profit - Private        3.26            1396
4                                     Physician        3.18              17
5                                        Tribal        3.00               1
6                            Government - State        2.86              36
7                            Government - Local        2.86             141
8   Government - Hospital District or Authority        2.82             172
9                          Government - Federal        2.75               8
10                                  Proprietary        2.70             476

### 📊 Query 6 Summary  
**Veterans Health Administration** hospitals have the highest average rating (**3.63**), followed by **Voluntary non-profit** institutions. In contrast, **Proprietary** and **Government-run** hospitals generally have lower average ratings. This suggests a potential relationship between ownership structure and overall hospital quality, which may reflect differences in resources, oversight, or priorities.

---
### 🧮 Query 7: Hospitals with 5-Star Rating by State

**Business Question**  
Which states have the highest number of top-rated (5-star) hospitals?

**Why this matters**  
This query identifies where the highest quality hospitals are geographically concentrated. It provides insight into regional healthcare excellence, helps benchmark state-level performance, and may guide patients seeking top-rated care.

**SQL Concepts Introduced**  
- `WHERE` clause with exact match filtering  
- Continued use of `GROUP BY`, `COUNT`, and `ORDER BY`

In [10]:
# Query 7: Count of hospitals with 5-star rating by state
query_7 = """
SELECT
    "State",
    COUNT(*) AS five_star_hospital_count
FROM
    hospital_info
WHERE
    "Hospital overall rating" = 5
GROUP BY
    "State"
ORDER BY
    five_star_hospital_count DESC;
"""

df_q7 = pd.read_sql_query(query_7, con=engine)
df_q7.head(10)  # Preview top 10 results

State  five_star_hospital_count
0    CA                        36
1    TX                        29
2    PA                        22
3    OH                        20
4    WI                        18
5    CO                        17
6    FL                        15
7    MI                        14
8    NC                        13
9    IN                        12

### 📊 Query 7 Summary  
**California (36)** and **Texas (29)** lead the nation in the number of 5-star rated hospitals, followed by **Pennsylvania (22)** and **Ohio (20)**. These high-performing states may serve as benchmarks for quality standards and best practices in healthcare service delivery. Recognizing regional excellence supports data-driven health policy and cross-state comparisons.

---
### 🧮 Query 8: Search Hospitals by Name (Keyword: "Children")

**Business Question**  
Which hospitals have the word “Children” in their name?

**Why this matters**  
Identifying hospitals with specialized care for children helps assess pediatric care availability. This is particularly important for location-based healthcare planning, funding, or policy development focused on child health.

**SQL Concepts Introduced**  
- Case-insensitive filtering using `ILIKE`
- Use of parameterized queries for dynamic filtering

In [11]:
# Query 8: Find hospitals with "Children" in the name (case-insensitive)
search_term = '%children%'

query_8 = """
SELECT
    "Facility ID",
    "Facility Name",
    "City/Town",
    "State"
FROM
    hospital_info
WHERE
    "Facility Name" ILIKE %s
LIMIT 20;
"""

df_q8 = pd.read_sql_query(query_8, con=engine, params=(search_term,))
df_q8

Facility ID                                      Facility Name  \
0       014015                         BEACON CHILDREN'S HOSPITAL   
1       013300                 THE CHILDREN'S HOSPITAL OF ALABAMA   
2       013301           USA HEALTH CHILDREN'S & WOMEN'S HOSPITAL   
3       033302                        PHOENIX CHILDREN'S HOSPITAL   
4       043300                       ARKANSAS CHILDREN'S HOSPITAL   
5       043301                 ARKANSAS CHILDREN'S NORTHWEST, INC   
6       050778          LOMA LINDA UNIVERSITY CHILDREN'S HOSPITAL   
7       053300                         VALLEY CHILDREN'S HOSPITAL   
8       053301           UCSF BENIOFF CHILDREN'S HOSPITAL OAKLAND   
9       053302                     CHILDREN'S HOSP OF LOS ANGELES   
10      053303               RADY CHILDREN'S HOSPITAL - SAN DIEGO   
11      053304               CHILDREN'S HOSPITAL OF ORANGE COUNTY   
12      053305   LUCILE SALTER PACKARD CHILDREN'S HSP AT STANFORD   
13      053306                     CHILDREN'S HOSPITAL AT MISSION   
14      053308          HEALTHBRIDGE CHILDREN'S HOSPITAL - ORANGE   
15      053309  MEMORIAL CARE MILLER CHILDREN'S &  WOMEN'S HOS...   
16      053311     SHRINERS HOSPITALS FOR CHILDREN NORTHERN CALIF   
17      063301                       CHILDREN'S HOSPITAL COLORADO   
18      063303    CHILDREN'S HOSPITAL COLORADO - COLORADO SPRINGS   
19      073300               CONNECTICUT CHILDRENS MEDICAL CENTER   

           City/Town State  
0            LUVERNE    AL  
1         BIRMINGHAM    AL  
2             MOBILE    AL  
3            PHOENIX    AZ  
4        LITTLE ROCK    AR  
5         SPRINGDALE    AR  
6         LOMA LINDA    CA  
7             MADERA    CA  
8            OAKLAND    CA  
9        LOS ANGELES    CA  
10         SAN DIEGO    CA  
11            ORANGE    CA  
12         PALO ALTO    CA  
13     MISSION VIEJO    CA  
14            ORANGE    CA  
15        LONG BEACH    CA  
16        SACRAMENTO    CA  
17            AURORA    CO  
18  COLORADO SPRINGS    CO  
19          HARTFORD    CT

### 📊 Query 8 Summary  
This query identified hospitals with **"Children"** in their facility names using a **case-insensitive search** (`ILIKE`). Results include institutions such as:

- **Beacon Children's Hospital** (AL)  
- **Phoenix Children's Hospital** (AZ)  
- **UCSF Benioff Children's Hospital Oakland** (CA)  
- **Connecticut Children's Medical Center** (CT)

These hospitals are likely specialized in pediatric care. Such filtering techniques are valuable when locating facilities based on naming patterns or keyword-based services in healthcare datasets.

---
### 🧮 Query 9: Average Hospital Rating by City (with ≥ 5 Hospitals)

**Business Question**  
Which cities have the highest average hospital ratings, considering only cities with at least five hospitals?

**Why this matters**  
Focusing on cities with multiple hospitals helps surface meaningful trends in hospital quality. Cities with consistently high ratings may reflect stronger local healthcare systems or competitive hospital environments.

**SQL Concepts Introduced**  
- `HAVING` clause to filter groups based on aggregated values
- Continued use of `GROUP BY`, `AVG`, and `ORDER BY`

In [12]:
# Query 9: Average hospital rating by city (only cities with ≥ 5 hospitals)
query_9 = """
SELECT 
    "City/Town" AS city,
    ROUND(AVG("Hospital overall rating"::NUMERIC), 2) AS avg_rating,
    COUNT(*) AS hospital_count
FROM 
    hospital_info
WHERE 
    "Hospital overall rating" IS NOT NULL
GROUP BY 
    "City/Town"
HAVING 
    COUNT(*) >= 5
ORDER BY 
    avg_rating DESC;
"""

df_q9 = pd.read_sql_query(query_9, con=engine)
df_q9.head(10)  # Preview top 10 cities with highest average rating

city  avg_rating  hospital_count
0       SAN DIEGO        4.40               5
1          AUSTIN        4.33               6
2         MADISON        4.20               5
3         WICHITA        4.20               5
4          BOSTON        4.17               6
5          AURORA        4.00               5
6  SALT LAKE CITY        4.00               5
7         LEBANON        3.89               9
8        PORTLAND        3.89               9
9           OMAHA        3.83               6

### 📊 Query 9 Summary  
**San Diego**, **Austin**, and **Madison** top the list of cities with the highest average hospital ratings, each maintaining an average above **4.2** across cities with **5 or more hospitals**. This suggests a strong standard of care in these urban healthcare systems. Limiting the analysis to cities with a minimum number of hospitals ensures more reliable and representative averages.

---
### 🧮 Query 10: Facility Names Shared Across States

**Business Question**  
Are there hospital facility names that appear in more than one U.S. state?

**Why this matters**  
Shared facility names across states may indicate chain hospitals, franchises, or naming conventions that could affect analysis. Identifying such patterns helps avoid potential confusion in state-level comparisons or geographic aggregations.

**SQL Concepts Introduced**  
- `COUNT(DISTINCT ...)` to measure unique occurrences  
- `HAVING` clause to filter aggregated results  
- Multi-column `ORDER BY` to break ties and improve readability

In [13]:
# Query 10: Count duplicate facility names across multiple states
query_10 = """
SELECT
    "Facility Name",
    COUNT(DISTINCT "State") AS state_count
FROM
    hospital_info
GROUP BY
    "Facility Name"
HAVING
    COUNT(DISTINCT "State") > 1
ORDER BY
    state_count DESC,
    "Facility Name"
"""

df_q10 = pd.read_sql_query(query_10, con=engine)
df_q10.head(10)  # Preview top 10 facility names found in multiple states

Facility Name  state_count
0                MEMORIAL HOSPITAL            7
1               COMMUNITY HOSPITAL            4
2      COMMUNITY MEMORIAL HOSPITAL            4
3              HOLY CROSS HOSPITAL            4
4          MEMORIAL MEDICAL CENTER            4
5             MERCY MEDICAL CENTER            4
6  SHRINERS HOSPITALS FOR CHILDREN            4
7         ST JOSEPH MEDICAL CENTER            4
8                ST LUKES HOSPITAL            4
9         ST MARY'S MEDICAL CENTER            4

### 📊 Query 10 Summary  
Several hospital names are shared across multiple states, suggesting the presence of chain systems or commonly used naming conventions. The most widely shared name is **MEMORIAL HOSPITAL**, found in **7 different states**, followed by **COMMUNITY HOSPITAL**, **HOLY CROSS HOSPITAL**, and others, each appearing in **4 states**. Recognizing these patterns helps ensure that state-level analyses do not mistakenly assume these facilities are unique to one location.

---
### 🧮 Query 11: Duplicate Facility Names Across States (via Self Join)

**Business Question**  
Which hospitals share the same name but are located in different U.S. states?

**Why this matters**  
Identifying facilities with identical names across states can reveal naming conventions, potential confusion in reporting, or affiliations between organizations. This is especially important when ensuring accurate identification and aggregation in healthcare analytics.

**SQL Concepts Introduced**  
- `JOIN` (self join on the same table)
- `DISTINCT` to remove duplicate row combinations
- `<>` for inequality filtering between states
- `ORDER BY` and `LIMIT` to sort and preview the results

In [14]:
# Query 11: Hospitals with the same name in different states (self join)
query_11 = """
SELECT DISTINCT
    h1."Facility Name" AS facility_name,
    h1."State" AS state_1,
    h2."State" AS state_2
FROM
    hospital_info h1
JOIN
    hospital_info h2
    ON h1."Facility Name" = h2."Facility Name"
   AND h1."State" <> h2."State"
ORDER BY
    facility_name
LIMIT 20;
"""

df_q11 = pd.read_sql_query(query_11, con=engine)  # Execute query
df_q11  # Preview first 20 duplicate facility names across states

facility_name state_1 state_2
0             BLUE MOUNTAIN HOSPITAL      UT      OR
1             BLUE MOUNTAIN HOSPITAL      OR      UT
2       BRIGHTWELL BEHAVIORAL HEALTH      MI      IN
3       BRIGHTWELL BEHAVIORAL HEALTH      IN      MI
4   CARROLL COUNTY MEMORIAL HOSPITAL      KY      MO
5   CARROLL COUNTY MEMORIAL HOSPITAL      MO      KY
6             CASCADE MEDICAL CENTER      WA      ID
7             CASCADE MEDICAL CENTER      ID      WA
8            CITIZENS MEDICAL CENTER      LA      TX
9            CITIZENS MEDICAL CENTER      LA      KS
10           CITIZENS MEDICAL CENTER      KS      LA
11           CITIZENS MEDICAL CENTER      TX      LA
12           CITIZENS MEDICAL CENTER      KS      TX
13           CITIZENS MEDICAL CENTER      TX      KS
14              CLAY COUNTY HOSPITAL      AL      IL
15              CLAY COUNTY HOSPITAL      IL      AL
16        COLUMBIA MEMORIAL HOSPITAL      NY      OR
17        COLUMBIA MEMORIAL HOSPITAL      OR      NY
18       COLUMBUS COMMUNITY HOSPITAL      TX      WI
19       COLUMBUS COMMUNITY HOSPITAL      WI      TX

### 📊 Query 11 Summary  
Several hospitals share the **same facility name across different states**, such as **"CITIZENS MEDICAL CENTER"** appearing in **Texas, Kansas, and Louisiana**, and **"BLUE MOUNTAIN HOSPITAL"** in **Utah** and **Oregon**. These duplicates often reflect national healthcare systems, regional naming patterns, or rebranded facilities. Identifying these is essential for accurate hospital-level analysis, especially when joining or aggregating by name alone.

---
### 🧮 Query 12: Duplicate Facility Names Within the Same State

**Business Question**  
Are there hospitals with the same facility name more than once within a single U.S. state?

**Why this matters**  
Duplicate hospital names within the same state may signal potential data entry errors, rebranded facilities, or healthcare systems with multiple branches. Identifying these duplicates is critical for avoiding data misinterpretation and ensuring clean, non-redundant records in downstream analysis.

**SQL Concepts Introduced**  
- `GROUP BY` on multiple columns
- `HAVING COUNT(*) > 1` for filtering group-level duplicates
- `ORDER BY` to prioritize most frequent duplicates

In [15]:
# Query 12: Hospitals with duplicate names within the same state
query_12 = """
SELECT
    "Facility Name" AS facility_name,
    "State",
    COUNT(*) AS facility_count
FROM
    hospital_info
GROUP BY
    "Facility Name", "State"
HAVING
    COUNT(*) > 1
ORDER BY
    facility_count DESC
"""

df_q12 = pd.read_sql_query(query_12, con=engine)
df_q12.head(10)  # Preview top 10 facility names duplicated within the same state

facility_name State  facility_count
0                         MEMORIAL HOSPITAL    TX               3
1                         MEMORIAL HOSPITAL    IL               3
2                         MEMORIAL HOSPITAL    OH               2
3  COMMUNITY HOSPITALS AND WELLNESS CENTERS    OH               2
4            CENTER FOR BEHAVIORAL MEDICINE    MO               2
5                   GOOD SAMARITAN HOSPITAL    CA               2
6               COMMUNITY MEMORIAL HOSPITAL    SD               2
7                        HOUSTON HEALTHCARE    GA               2
8                   MENTAL HEALTH INSTITUTE    IA               2
9               MEMORIAL HOSPITAL AND MANOR    GA               2

### 📊 Query 12 Summary  
Several hospitals share the same facility name **within the same state**, which may indicate **duplicate entries**, **multiple branches under the same name**, or **data inconsistencies**. For example, **"MEMORIAL HOSPITAL"** appears **3 times in Texas** and **3 times in Illinois**, suggesting this name is particularly common and should be handled carefully in entity-level analysis or deduplication steps.

---
### 🧮 Query 13: States with Above-National-Average Hospital Ratings

**Business Question**  
Which states have an average hospital rating that exceeds the national average?

**Why this matters**  
Identifying states that outperform the national benchmark can help pinpoint regions with higher quality of care or better healthcare infrastructure. This analysis also sets the stage for further benchmarking and comparative performance evaluations.

**SQL Concepts Introduced**  
- Subquery in `HAVING` clause for dynamic filtering
- Type casting with `::FLOAT` to ensure accurate averages
- Continued use of `GROUP BY`, `AVG`, and `ORDER BY`

In [16]:
# Query 13: States with above-national-average hospital ratings
query_13 = """
SELECT
    "State",
    ROUND(AVG("Hospital overall rating"::FLOAT)::NUMERIC, 2) AS avg_state_rating
FROM
    hospital_info
WHERE
    "Hospital overall rating" IS NOT NULL
GROUP BY
    "State"
HAVING
    AVG("Hospital overall rating"::FLOAT) > (
        SELECT
            AVG("Hospital overall rating"::FLOAT)
        FROM
            hospital_info
        WHERE
            "Hospital overall rating" IS NOT NULL
    )
ORDER BY
    avg_state_rating DESC
"""

df_q13 = pd.read_sql_query(query_13, con=engine)
df_q13.head(10)  # Preview top 10 states with above-average ratings

State  avg_state_rating
0    UT              4.29
1    SD              4.23
2    CO              4.02
3    MN              3.95
4    WI              3.89
5    ID              3.88
6    MT              3.81
7    OH              3.61
8    DE              3.57
9    NE              3.56

### 📊 Query 13 Summary  
States like **Utah (4.29)**, **South Dakota (4.23)**, and **Colorado (4.02)** have the highest average hospital ratings — significantly above the national average. These results help surface top-performing states and can guide benchmark comparisons or quality improvement programs in other regions.

---
### 🧮 Query 14: Top-Rated Hospitals in Each State

**Business Question**  
Which hospitals are the highest-rated in each U.S. state?

**Why this matters**  
This query highlights the top-performing hospitals by quality rating across each state. It provides insights into centers of excellence and helps identify benchmarks for hospital quality at the state level.

**SQL Concepts Introduced**  
- `RANK()` window function to rank hospitals within each state  
- `PARTITION BY` and `ORDER BY` within a window function  
- `WHERE IS NOT NULL` to filter valid ratings

In [17]:
# Query 14: Top-rated hospitals in each state using RANK()
query_14 = """
SELECT 
    "Facility ID",
    "Facility Name",
    "City/Town",
    "State",
    "Hospital overall rating",
    RANK() OVER (
        PARTITION BY "State"
        ORDER BY "Hospital overall rating"::FLOAT DESC
    ) AS rating_rank
FROM 
    hospital_info
WHERE 
    "Hospital overall rating" IS NOT NULL
"""

df_q14 = pd.read_sql_query(query_14, con=engine)

# Filter only top-ranked (rank = 1) hospitals per state
df_q14_top = df_q14[df_q14["rating_rank"] == 1].sort_values(by=["State", "Facility Name"])
df_q14_top.head(10)

Facility ID                                      Facility Name  \
0       020008                         BARTLETT REGIONAL HOSPITAL   
7       010083                                     BALDWIN HEALTH   
9       010125                        LAKELAND COMMUNITY HOSPITAL   
8       01019F  VA CENTRAL ALABAMA HEALTHCARE SYSTEM - MONTGOMERY   
58      04009F                  FAYETTEVILLE AR VA MEDICAL CENTER   
59      040010                  MERCY HOSPITAL NORTHWEST ARKANSAS   
95      031318                       BANNER PAYSON MEDICAL CENTER   
99      030123       HONORHEALTH SCOTTSDALE THOMPSON PEAK MED CTR   
98      030103                               MAYO CLINIC HOSPITAL   
97      03033F              VA NORTHERN ARIZONA HEALTHCARE SYSTEM   

       City/Town State  Hospital overall rating  rating_rank  
0         JUNEAU    AK                      4.0            1  
7          FOLEY    AL                      5.0            1  
9     HALEYVILLE    AL                      5.0            1  
8     MONTGOMERY    AL                      5.0            1  
58  FAYETTEVILLE    AR                      5.0            1  
59        ROGERS    AR                      5.0            1  
95        PAYSON    AZ                      5.0            1  
99    SCOTTSDALE    AZ                      5.0            1  
98       PHOENIX    AZ                      5.0            1  
97      PRESCOTT    AZ                      5.0            1

### 📊 Query 14 Summary  
Several hospitals in each state have achieved the highest possible overall rating. For example, **Alabama (AL)** has multiple top-rated hospitals including **BALDWIN HEALTH**, **LAKELAND COMMUNITY HOSPITAL**, and **VA CENTRAL ALABAMA HEALTHCARE SYSTEM – MONTGOMERY**, all with a **5-star rating**. Similarly, **Arizona (AZ)** features facilities like **MAYO CLINIC HOSPITAL** and **HONORHEALTH SCOTTSDALE THOMPSON PEAK MED CTR** at the top tier. This query showcases **centers of excellence** across the nation and introduces advanced SQL ranking techniques.

---
### 🧮 Query 15: Hospitals in the 10 Most Common Cities

**Business Question**  
Which hospitals are located in the cities that appear most frequently in the dataset?

**Why this matters**  
Some cities may have multiple healthcare facilities due to higher population density or their role as regional hubs. Identifying hospitals in these high-density cities provides insight into urban healthcare infrastructure and distribution.

**SQL Concepts Introduced**  
- Subqueries with `IN`  
- `GROUP BY` and `ORDER BY` inside subquery  
- Use of `DISTINCT` for deduplication

In [18]:
# Query 15: Hospitals in the 10 most common cities
query_15 = """
SELECT DISTINCT 
    "Facility ID", 
    "Facility Name", 
    "City/Town", 
    "State"
FROM 
    hospital_info
WHERE 
    "City/Town" IN (
        SELECT 
            "City/Town"
        FROM 
            hospital_info
        GROUP BY 
            "City/Town"
        ORDER BY 
            COUNT(*) DESC
        LIMIT 10
    )
ORDER BY 
    "City/Town", 
    "Facility Name"
"""

df_q15 = pd.read_sql_query(query_15, con=engine)
df_q15.head(10)  # Preview top 10 hospitals from the 10 most common cities

Facility ID                                      Facility Name City/Town  \
0      140182           ADVOCATE ILLINOIS MASONIC MEDICAL CENTER   CHICAGO   
1      140048                          ADVOCATE TRINITY HOSPITAL   CHICAGO   
2      140117           AMITA HEALTH RESURRECTION MEDICAL CENTER   CHICAGO   
3      143300  ANN & ROBERT H LURIE CHILDRENS HOSPITAL OF CHI...   CHICAGO   
4      144010                  CHICAGO READ MENTAL HEALTH CENTER   CHICAGO   
5      140251                     COMMUNITY FIRST MEDICAL CENTER   CHICAGO   
6      144039                             GARFIELD PARK HOSPITAL   CHICAGO   
7      144026                                 HARTGROVE HOSPITAL   CHICAGO   
8      140133                                HOLY CROSS HOSPITAL   CHICAGO   
9      140206                               HUMBOLDT PARK HEALTH   CHICAGO   

  State  
0    IL  
1    IL  
2    IL  
3    IL  
4    IL  
5    IL  
6    IL  
7    IL  
8    IL  
9    IL

### 📊 Query 15 Summary  
**Chicago (IL)** stands out as the most common city for hospital facilities, featuring a wide range of institutions including **community hospitals**, **children’s hospitals**, and **mental health centers**. Analyzing hospitals in these high-density cities helps illuminate patterns of healthcare availability and specialization in major urban areas.

---
### 🧮 Query 16: Hospitals Grouped into 4 Performance Tiers

**Business Question**  
How can we categorize hospitals into four performance-based tiers using overall rating data?

**Why this matters**  
Segmenting hospitals into quartiles provides a simplified and structured way to benchmark hospital performance across the U.S. This is especially useful for reporting, policy decisions, and identifying areas needing improvement or celebrating excellence.

**SQL Concepts Introduced**  
- `NTILE()` window function to split rows into equal-sized ranking groups  
- Continued use of `ORDER BY`, casting for numeric operations, and filtering with `WHERE IS NOT NULL`

In [19]:
# Query 16: Group hospitals into performance quartiles using NTILE
query_16 = """
SELECT
    "Facility ID",
    "Facility Name",
    "City/Town",
    "State",
    "Hospital overall rating",
    NTILE(4) OVER (
        ORDER BY "Hospital overall rating"::INT DESC
    ) AS performance_quartile
FROM
    hospital_info
WHERE
    "Hospital overall rating" IS NOT NULL
ORDER BY
    performance_quartile,
    "Hospital overall rating"::INT DESC
LIMIT
    20;
"""

df_q16 = pd.read_sql_query(query_16, con=engine)
df_q16  # Display sample hospitals with performance quartiles

Facility ID                                      Facility Name  \
0       490050                           VIRGINIA HOSPITAL CENTER   
1       500124                     EVERGREENHEALTH MEDICAL CENTER   
2       420089                         EAST COOPER MEDICAL CENTER   
3       220119              BRIGHAM AND WOMEN'S FAULKNER HOSPITAL   
4       230097                              MUNSON MEDICAL CENTER   
5       47001F                  WHITE RIVER JCT VA MEDICAL CENTER   
6       140010  NORTHSHORE UNIVERSITY HEALTHSYSTEM - EVANSTON ...   
7       201302                                      LINCOLNHEALTH   
8       45086F  TEMPLE VA MEDICAL CENTER (VA CENTRAL TEXAS HEA...   
9       450867                          ASCENSION SETON NORTHWEST   
10      520091                        HOWARD YOUNG MEDICAL CENTER   
11      050013                         ADVENTIST HEALTH ST HELENA   
12      530015                            ST JOHNS MEDICAL CENTER   
13      060028                              SAINT JOSEPH HOSPITAL   
14      50011F                          SPOKANE VA MEDICAL CENTER   
15      521357           MERCY WALWORTH HOSPITAL & MEDICAL CENTER   
16      100043                             MEASE DUNEDIN HOSPITAL   
17      450851  BAYLOR SCOTT & WHITE HEART & VASCULAR HOSPITAL...   
18      420065             BON SECOURS-ST FRANCIS XAVIER HOSPITAL   
19      450893                   THE HEART HOSPITAL BAYLOR DENTON   

               City/Town State  Hospital overall rating  performance_quartile  
0              ARLINGTON    VA                      5.0                     1  
1               KIRKLAND    WA                      5.0                     1  
2         MOUNT PLEASANT    SC                      5.0                     1  
3          JAMAICA PLAIN    MA                      5.0                     1  
4          TRAVERSE CITY    MI                      5.0                     1  
5   WHITE RIVER JUNCTION    VT                      5.0                     1  
6               EVANSTON    IL                      5.0                     1  
7           DAMARISCOTTA    ME                      5.0                     1  
8                 TEMPLE    TX                      5.0                     1  
9                 AUSTIN    TX                      5.0                     1  
10              WOODRUFF    WI                      5.0                     1  
11          SAINT HELENA    CA                      5.0                     1  
12               JACKSON    WY                      5.0                     1  
13                DENVER    CO                      5.0                     1  
14               SPOKANE    WA                      5.0                     1  
15           LAKE GENEVA    WI                      5.0                     1  
16               DUNEDIN    FL                      5.0                     1  
17                DALLAS    TX                      5.0                     1  
18            CHARLESTON    SC                      5.0                     1  
19                DENTON    TX                      5.0                     1

### 📊 Query 16 Summary

The output lists hospitals grouped into four performance quartiles based on their overall rating. The **top quartile (performance_quartile = 1)** includes hospitals with the highest ratings, such as **Virginia Hospital Center (ARLINGTON, VA)** and **EvergreenHealth Medical Center (KIRKLAND, WA)**, both with a perfect score of 5.0.

This segmentation offers a clear way to benchmark hospital quality across states, useful for policymakers, healthcare providers, and researchers aiming to identify high and low performers.

---
### 🧮 Query 17: States with the Most 1-Star Hospitals Using a CTE

**Business Question**  
Which states have the highest number of hospitals rated with only one star?

**Why this matters**  
Identifying states with a concentration of underperforming hospitals helps focus improvement efforts and resource allocation to areas where patient care may be most at risk.

**SQL Concepts Introduced**  
- Common Table Expressions (`WITH` clause) for modular query design  
- Filtering with `WHERE`  
- Aggregation using `COUNT`, `GROUP BY`, and sorting with `ORDER BY`
- Limiting results with `LIMIT`

In [20]:
# Query 17: States with the most 1-star hospitals using a CTE
query_17 = """
WITH underperforming_hospitals AS (
    SELECT *
    FROM hospital_info
    WHERE "Hospital overall rating" = '1'
)
SELECT
    "State",
    COUNT(*) AS one_star_count
FROM
    underperforming_hospitals
GROUP BY
    "State"
ORDER BY
    one_star_count DESC
LIMIT 10;
"""

df_q17 = pd.read_sql_query(query_17, con=engine)
df_q17

State  one_star_count
0    FL              34
1    NY              33
2    CA              31
3    IL              14
4    NJ              12
5    KY              12
6    GA              11
7    NC               9
8    WA               9
9    PA               8

### 📊 Query 17 Summary

The output reveals that **Florida (FL)**, **New York (NY)**, and **California (CA)** have the highest counts of 1-star hospitals, with Florida leading at **34** such hospitals. These states may require targeted quality improvement initiatives to enhance patient outcomes. This query highlights the use of CTEs to simplify complex filtering and aggregation tasks.

---
### 🧮 Query 18: Top-Rated Hospitals Within Each City

**Business Question**  
Which hospitals have the highest overall rating in their respective cities?

**Why this matters**  
This query surfaces the top-performing hospitals at the city level, enabling localized quality benchmarking. It demonstrates the use of correlated subqueries to compare each hospital's rating against the maximum rating within its city, a common pattern in comparative analytics.

**SQL Concepts Introduced**  
- Correlated subqueries  
- Filtering with aggregate functions in subqueries  
- Ordering results  

In [21]:
# Query 18: Hospitals with the highest overall rating in their city using a correlated subquery
query_18 = """
SELECT 
    h."Facility ID", 
    h."Facility Name", 
    h."City/Town", 
    h."State", 
    h."Hospital overall rating"
FROM 
    hospital_info h
WHERE 
    h."Hospital overall rating" IS NOT NULL
    AND h."Hospital overall rating" = (
        SELECT MAX(h2."Hospital overall rating")
        FROM hospital_info h2
        WHERE h2."City/Town" = h."City/Town"
    )
ORDER BY 
    h."City/Town" ASC
LIMIT 20
"""

df_q18 = pd.read_sql_query(query_18, con=engine)
df_q18

Facility ID                                      Facility Name   City/Town  \
0       190034                         ABBEVILLE GENERAL HOSPITAL   ABBEVILLE   
1       430014                                     AVERA ST LUKES    ABERDEEN   
2       450229                            HENDRICK MEDICAL CENTER     ABILENE   
3       490053                         JOHNSTON MEMORIAL HOSPITAL    ABINGDON   
4       390231                        JEFFERSON ABINGTON HOSPITAL    ABINGTON   
5       370180                    CHICKASAW NATION MEDICAL CENTER         ADA   
6       370020                                 MERCY HOSPITAL ADA         ADA   
7       670073                     METHODIST HOSPITAL FOR SURGERY     ADDISON   
8       230005    PROMEDICA CHARLES AND VIRGINIA HICKMAN HOSPITAL      ADRIAN   
9       340099                     VIDANT ROANOKE CHOWAN HOSPITAL     AHOSKIE   
10      120026                           PALI MOMI MEDICAL CENTER        AIEA   
11      420082                      AIKEN REGIONAL MEDICAL CENTER       AIKEN   
12      360351                  CRYSTAL CLINIC ORTHOPAEDIC CENTER       AKRON   
13      010016                      SHELBY BAPTIST MEDICAL CENTER   ALABASTER   
14      050211                                   ALAMEDA HOSPITAL     ALAMEDA   
15      320004                       CHRISTUS SOUTHERN NEW MEXICO  ALAMOGORDO   
16      060008           SAN LUIS VALLEY REGIONAL  MEDICAL CENTER     ALAMOSA   
17      330057                                ST PETER'S HOSPITAL      ALBANY   
18      340119                     STANLY REGIONAL MEDICAL CENTER   ALBEMARLE   
19      240043  MAYO CLINIC HEALTH SYSTEM - ALBERT LEA AND AUSTIN  ALBERT LEA   

   State  Hospital overall rating  
0     LA                      3.0  
1     SD                      5.0  
2     TX                      3.0  
3     VA                      2.0  
4     PA                      2.0  
5     OK                      5.0  
6     OK                      5.0  
7     TX                      4.0  
8     MI                      4.0  
9     NC                      2.0  
10    HI                      4.0  
11    SC                      2.0  
12    OH                      5.0  
13    AL                      3.0  
14    CA                      4.0  
15    NM                      2.0  
16    CO                      1.0  
17    NY                      4.0  
18    NC                      4.0  
19    MN                      5.0

### 📊 Query 18 Summary

The output lists hospitals that received the **highest overall rating within their respective cities**. For example, both **Chickasaw Nation Medical Center** and **Mercy Hospital Ada** in *Ada, OK* have top ratings of **5.0**. This demonstrates how correlated subqueries allow filtering rows based on group-level maximum values, enabling city-level benchmarking of hospital quality.

This technique is useful when identifying local leaders or top performers without pre-aggregating data.

---
### 🧮 Query 19: Most Recent Facility in Each City

**Business Question**  
Which is the most recently added hospital facility in each city based on the highest Facility ID?

**Why this matters**  
This query uses window functions (`ROW_NUMBER()`) to rank hospitals within each city by Facility ID, highlighting the latest facilities. Identifying recent entries can help understand healthcare expansion or infrastructure development patterns at a local level.

**SQL Concepts Introduced**  
- Use of `WITH` Common Table Expressions (CTEs) for layered query building  
- `ROW_NUMBER()` window function to rank rows within partitions  
- Filtering on window function results to extract top-ranked rows per group

In [22]:
# Query 19: Most Recent Facility in Each City using ROW_NUMBER()
query_19 = """
WITH ranked_hospitals AS (
    SELECT *,
           ROW_NUMBER() OVER (
               PARTITION BY "City/Town"
               ORDER BY "Facility ID" DESC
           ) AS row_num
    FROM
        hospital_info
)
SELECT
    "Facility ID", 
    "Facility Name", 
    "City/Town", 
    "State", 
    "Hospital overall rating"
FROM
    ranked_hospitals
WHERE
    row_num = 1
ORDER BY
    "City/Town"
LIMIT 20
"""

df_q19 = pd.read_sql_query(query_19, con=engine)
df_q19.head(10)

Facility ID                                      Facility Name  City/Town  \
0      421301                      ABBEVILLE AREA MEDICAL CENTER  ABBEVILLE   
1      500031                    GRAYS HARBOR COMMUNITY HOSPITAL   ABERDEEN   
2      454122              OCEANS BEHAVIORAL HOSPITAL OF ABILENE    ABILENE   
3      490053                         JOHNSTON MEMORIAL HOSPITAL   ABINGDON   
4      390231                        JEFFERSON ABINGTON HOSPITAL   ABINGTON   
5      251334                    CHOCTAW REGIONAL MEDICAL CENTER   ACKERMAN   
6      374016                        ROLLING HILLS HOSPITAL, LLC        ADA   
7      670073                     METHODIST HOSPITAL FOR SURGERY    ADDISON   
8      110101  COOK  MEDICAL CENTER  A CAMPUS OF TIFT REG MED...       ADEL   
9      230005    PROMEDICA CHARLES AND VIRGINIA HICKMAN HOSPITAL     ADRIAN   

  State  Hospital overall rating  
0    SC                      NaN  
1    WA                      2.0  
2    TX                      NaN  
3    VA                      2.0  
4    PA                      2.0  
5    MS                      NaN  
6    OK                      NaN  
7    TX                      4.0  
8    GA                      NaN  
9    MI                      4.0

### 📊 Query 19 Summary

This output lists the **most recent facility in each city**, determined by the highest `"Facility ID"` value.  
Note that some facilities do not have an overall rating (`NaN`), which may indicate new or unassessed hospitals.  
This query demonstrates the use of **window functions** (`ROW_NUMBER()`) to rank and filter records within groups for granular insights.

---
### 🧮 Query 20: Cities with the Widest Range in Hospital Ratings

**Business Question**  
Which cities show the largest variation between their highest and lowest hospital overall ratings?

**Why this matters**  
Wide rating ranges may indicate uneven quality of care within a city, highlighting areas for potential improvement or more detailed investigation.

**SQL Concepts Introduced**  
- Use of aggregate functions (`MAX`, `MIN`, `COUNT`)
- Calculating differences between aggregates for derived metrics
- Grouping by multiple columns with `GROUP BY`
- Filtering groups using `HAVING`
- Sorting results with multiple criteria in `ORDER BY`
- Casting data types using `CAST`

In [23]:
# Query 20: Find cities with the widest range in hospital ratings
query_20 = """
SELECT
    "City/Town",
    "State",
    MAX(CAST("Hospital overall rating" AS FLOAT)) AS max_rating,
    MIN(CAST("Hospital overall rating" AS FLOAT)) AS min_rating,
    COUNT(*) AS hospital_count,
    MAX(CAST("Hospital overall rating" AS FLOAT)) - MIN(CAST("Hospital overall rating" AS FLOAT)) AS rating_range
FROM
    hospital_info
WHERE
    "Hospital overall rating" IS NOT NULL
GROUP BY
    "City/Town",
    "State"
HAVING
    COUNT(*) >= 2
ORDER BY
    rating_range DESC,
    hospital_count DESC
LIMIT 10;
"""

df_q20 = pd.read_sql_query(query_20, con=engine)
df_q20

City/Town State  max_rating  min_rating  hospital_count  rating_range
0        CHICAGO    IL         5.0         1.0              21           4.0
1    LOS ANGELES    CA         5.0         1.0              14           4.0
2       NEW YORK    NY         5.0         1.0              11           4.0
3        PHOENIX    AZ         5.0         1.0              11           4.0
4   PHILADELPHIA    PA         5.0         1.0              11           4.0
5         DALLAS    TX         5.0         1.0              10           4.0
6  OKLAHOMA CITY    OK         5.0         1.0               9           4.0
7          MIAMI    FL         5.0         1.0               8           4.0
8    SAINT LOUIS    MO         5.0         1.0               7           4.0
9        SEATTLE    WA         5.0         1.0               6           4.0

### 📊 Query 20 Summary

Cities like **Chicago, IL**, **Los Angeles, CA**, and **New York, NY** exhibit the widest range in hospital overall ratings, from **1.0** up to the maximum of **5.0**. This indicates significant variability in hospital quality within these urban centers. Understanding such disparities is crucial for targeted healthcare improvements and resource allocation.

---
### 🧮 Query 21: Simulated Filtering of Facilities Joined After 2015

**Business Question**  
Which hospitals joined the dataset after the year 2015?

**Why this matters**  
Filtering by join date simulates temporal analysis commonly used in healthcare to track recent additions or changes to facilities. It helps understand growth trends and identify new healthcare providers entering the market.

**SQL Concepts Introduced**  
- Loading full dataset for post-processing in Python
- Filtering data based on date values (simulated here since original dataset lacks join dates) 

In [24]:
# Query 21: Simulated Filtering of Facilities Joined After 2015
# Load full hospital data and simulate join dates
df_full = pd.read_sql_query("SELECT * FROM hospital_info", con=engine)

# Generate random 'Join Date' between 2000 and 2022 for demonstration
np.random.seed(42)
random_dates = pd.to_datetime(
    np.random.choice(
        pd.date_range(start="2000-01-01", end="2022-12-31"),
        size=len(df_full)
    )
)
df_full["Join Date"] = random_dates

# Filter hospitals joined after 2015-12-31
df_q21 = df_full[df_full["Join Date"] > "2015-12-31"]

# Show sample output sorted by Join Date
df_q21[["Facility ID", "Facility Name", "State", "Join Date"]].sort_values("Join Date").head(10)

Facility ID                                 Facility Name State  \
2591      241378                        NEW ULM MEDICAL CENTER    MN   
3742      370026             ST MARY'S REGIONAL MEDICAL CENTER    OK   
276       040154         BAPTIST HEALTH MEDICAL CENTER- CONWAY    AR   
157       030089             BANNER THUNDERBIRD MEDICAL CENTER    AZ   
4191      420026  MUSC HEALTH COLUMBIA MEDICAL CENTER DOWNTOWN    SC   
132       030014   HONOR HEALTH JOHN C. LINCOLN MEDICAL CENTER    AZ   
3588      36008F                 CHILLICOTHE VA MEDICAL CENTER    OH   
2550      241337                    NORTH VALLEY HEALTH CENTER    MN   
4386      440218      SAINT THOMAS HOSPITAL FOR SPINAL SURGERY    TN   
3816      371313                MERCY HOSPITAL KINGFISHER, INC    OK   

      Join Date  
2591 2016-01-02  
3742 2016-01-07  
276  2016-01-11  
157  2016-01-12  
4191 2016-01-13  
132  2016-01-21  
3588 2016-01-23  
2550 2016-01-24  
4386 2016-01-24  
3816 2016-01-26

### 📊 Query 21 Summary
The output shows hospitals that have "joined" the dataset after December 31, 2015, based on simulated join dates. For example, **NEW ULM MEDICAL CENTER (MN)** and **ST MARY'S REGIONAL MEDICAL CENTER (OK)** are among the earliest new entries in 2016. This filtering helps simulate how recent facility additions can be tracked over time, which is valuable for monitoring healthcare market growth and facility updates.

---
### 🧮 Query 22: Analyze Name Patterns Using SQL String Functions

**Business Question**  
How can we analyze hospital name patterns using string functions, such as extracting prefixes, measuring name length, and locating keywords?

**Why this matters**  
Understanding naming conventions and patterns can help in data standardization, entity resolution, and text-based categorization, which are common tasks in data cleaning and feature engineering.

**SQL Concepts Introduced**  
- Use of string functions: `LEFT()`, `CHAR_LENGTH()`, `POSITION()`  
- Filtering rows based on string pattern positions  
- Ordering results by computed string length  

In [25]:
# Query 22: Analyze Name Patterns Using SQL String Functions
query_22 = """
SELECT
    "Facility Name",
    LEFT("Facility Name", 5) AS name_prefix,
    CHAR_LENGTH("Facility Name") AS name_length,
    POSITION('HOSPITAL' IN UPPER("Facility Name")) AS hospital_position
FROM
    hospital_info
WHERE
    POSITION('HOSPITAL' IN UPPER("Facility Name")) > 0
ORDER BY
    name_length DESC
LIMIT 10
"""

df_q22 = pd.read_sql_query(query_22, con=engine)
df_q22

Facility Name name_prefix  name_length  \
0  THE HEALTH CARE AUTHORITY OF THE CITY OF GREEN...       THE H           72   
1  UT OF TEXAS  SOUTHWESTERN UNIVERSITY HOSPITAL ...       UT OF           71   
2  UOFL HEALTH - JEWISH HOSPITAL and Mary & Eliza...       UOFL            59   
3  AERIES HEALTHCARE OF ILLINOIS, INC., DBA RIVER...       AERIE           59   
4  BAYLOR SCOTT & WHITE HEART & VASCULAR HOSPITAL...       BAYLO           55   
5  NORTHSHORE UNIVERSITY HEALTHSYSTEM - EVANSTON ...       NORTH           54   
6  ANNA HOSPITAL CORPORATION D/B/A UNION COUNTY H...       ANNA            53   
7  UCSF LANGLEY PORTER PSYCHIATRIC HOSPITAL AND C...       UCSF            52   
8  KAISER FOUNDATION HOSPITAL - ORANGE COUNTY - A...       KAISE           52   
9  ROBERT WOOD JOHNSON UNIVERSITY HOSPITAL AT HAM...       ROBER           51   

   hospital_position  
0                 65  
1                 38  
2                 22  
3                 52  
4                 39  
5                 47  
6                  6  
7                 33  
8                 19  
9                 32

### 📊 Query 22 Summary
The analysis reveals patterns in hospital names, such as common prefixes like "**THE H**" or "**UT OF**," and varying name lengths ranging from **51** to **72** characters. The keyword "HOSPITAL" appears at different positions within the names, indicating diversity in naming conventions. These insights can assist in data standardization and text processing efforts.

---
### 🧮 Query 23: Compare 5-Star Hospitals Offering Emergency Services Using EXCEPT and UNION

**Business Question**  
Which hospitals have a 5-star rating but do **not** offer emergency services?

**Why this matters**  
This query identifies highly rated hospitals that may lack critical emergency services, highlighting potential gaps in care accessibility despite high overall quality.

**SQL Concepts Introduced**  
- Use of set operations: `EXCEPT` to find differences between two result sets  
- Logical filtering with `WHERE` clause  
- Combining multiple SELECT queries

In [26]:
# Query 23: Find 5-star hospitals that do not offer emergency services using EXCEPT
query_23 = """
(
    SELECT
        "Facility ID",
        "Facility Name",
        "City/Town",
        "State"
    FROM
        hospital_info
    WHERE
        "Hospital overall rating" = 5
)
EXCEPT
(
    SELECT
        "Facility ID",
        "Facility Name",
        "City/Town",
        "State"
    FROM
        hospital_info
    WHERE
        "Hospital overall rating" = 5
        AND "Emergency Services" = 'Yes'
)
"""

df_q23 = pd.read_sql_query(query_23, con=engine)
df_q23

Facility ID                                     Facility Name  \
0       031318                      BANNER PAYSON MEDICAL CENTER   
1       420102                  PRISMA HEALTH  PATEWOOD HOSPITAL   
2       390332                    GEISINGER ST. LUKE'S  HOSPITAL   
3       050696                              KECK HOSPITAL OF USC   
4       340148               NOVANT HEALTH MEDICAL PARK HOSPITAL   
5       360071                          VAN WERT COUNTY HOSPITAL   
6       360041                  PARMA COMMUNITY GENERAL HOSPITAL   
7       220088                      NEW ENGLAND BAPTIST HOSPITAL   
8       050424                            SCRIPPS GREEN HOSPITAL   
9       030138                   WESTERN REGIONAL MEDICAL CENTER   
10      230133        MUNSON HEALTHCARE OTSEGO MEMORIAL HOSPITAL   
11      310005                          HUNTERDON MEDICAL CENTER   
12      240076                                  BUFFALO HOSPITAL   
13      440091                   MEMORIAL HEALTHCARE SYSTEM, INC   
14      26007F                    POPLAR BLUFF VA MEDICAL CENTER   
15      490101                          INOVA FAIR OAKS HOSPITAL   
16      670071  TEXAS HEALTH HEART & VASCULAR HOSPITAL ARLINGTON   
17      330205                     ST ANTHONY COMMUNITY HOSPITAL   
18      390326              ST LUKE'S HOSPITAL - ANDERSON CAMPUS   
19      360352                   SURGICAL HOSPITAL AT SOUTHWOODS   
20      050013                        ADVENTIST HEALTH ST HELENA   
21      110233         SOUTHEASTERN REGIONAL MEDICAL CENTER, INC   
22      360351                 CRYSTAL CLINIC ORTHOPAEDIC CENTER   

        City/Town State  
0          PAYSON    AZ  
1      GREENVILLE    SC  
2     ORWIGSBGURG    PA  
3     LOS ANGELES    CA  
4   WINSTON-SALEM    NC  
5        VAN WERT    OH  
6           PARMA    OH  
7          BOSTON    MA  
8        LA JOLLA    CA  
9        GOODYEAR    AZ  
10        GAYLORD    MI  
11     FLEMINGTON    NJ  
12        BUFFALO    MN  
13    CHATTANOOGA    TN  
14   POPLAR BLUFF    MO  
15        FAIRFAX    VA  
16      ARLINGTON    TX  
17        WARWICK    NY  
18         EASTON    PA  
19     YOUNGSTOWN    OH  
20   SAINT HELENA    CA  
21         NEWNAN    GA  
22          AKRON    OH

### 📊 Query 23 Summary

This query identifies **5-star hospitals that do not offer emergency services**. It highlights facilities that achieve the highest rating for overall quality but lack emergency service capabilities, which might be important for certain stakeholders evaluating hospital readiness and scope of care. The use of the `EXCEPT` set operation effectively filters out hospitals that do offer emergency services, demonstrating how SQL can be used for set-based comparisons in real-world scenarios.

---
### 🧮 Query 24: Pivot Hospital Ratings into Columns

**Business Question**  
How are hospital overall ratings distributed across the dataset, shown as counts per rating category?

**Why this matters**  
This pivot-style aggregation transforms rating counts into a wide format, providing a clear snapshot of rating distribution. It aids stakeholders in quickly assessing quality spread and identifying potential areas for improvement.

**SQL Concepts Introduced**  
- Conditional aggregation using `CASE WHEN` inside `SUM`
- Pivoting data from rows to columns without explicit pivot functions

In [27]:
# Query 24: Pivot hospital ratings into columns
query_24 = """
SELECT
    COUNT(*) AS total_hospitals,
    SUM(CASE WHEN "Hospital overall rating" = 1 THEN 1 ELSE 0 END) AS rating_1,
    SUM(CASE WHEN "Hospital overall rating" = 2 THEN 1 ELSE 0 END) AS rating_2,
    SUM(CASE WHEN "Hospital overall rating" = 3 THEN 1 ELSE 0 END) AS rating_3,
    SUM(CASE WHEN "Hospital overall rating" = 4 THEN 1 ELSE 0 END) AS rating_4,
    SUM(CASE WHEN "Hospital overall rating" = 5 THEN 1 ELSE 0 END) AS rating_5
FROM
    hospital_info
WHERE
    "Hospital overall rating" IS NOT NULL
"""

df_q24 = pd.read_sql_query(query_24, con=engine)
df_q24

total_hospitals  rating_1  rating_2  rating_3  rating_4  rating_5
0             2812       275       583       816       758       380

### 📊 Query 24 Summary

The pivot table shows the distribution of hospital overall ratings across the dataset. Out of **2,812** hospitals with available ratings, the largest group falls into the **3-star** category with **816** hospitals, followed closely by 4-star and 2-star ratings. This pivot-style aggregation helps visualize the spread of quality ratings and identify concentration trends in hospital performance.

---
### 🧮 Query 25: Top States by Average Hospital Rating (Cities with ≥ 3 Hospitals)

**Business Question**  
Which states have the highest average hospital ratings, considering only cities with at least three hospitals?

**Why this matters**  
This nested CTE query provides a layered analysis by first calculating city-level averages for hospitals meeting a minimum count threshold, then aggregating these to the state level. It highlights states with generally higher quality care in cities of significant hospital presence, filtering out small or outlier locations.

**SQL Concepts Introduced**  
- Nested Common Table Expressions (CTEs)  
- Aggregation and filtering with `HAVING`  
- Use of `ROUND` for numeric formatting  
- Layered aggregation from city to state level  

In [28]:
# Query 25: Use nested CTEs to find top states by average rating (only cities with ≥ 3 hospitals)
query_25 = """
WITH city_level AS (
    SELECT
        "City/Town",
        "State",
        AVG("Hospital overall rating"::FLOAT) AS avg_city_rating,
        COUNT(*) AS hospital_count
    FROM
        hospital_info
    WHERE
        "Hospital overall rating" IS NOT NULL
    GROUP BY
        "City/Town",
        "State"
    HAVING
        COUNT(*) >= 3
),
state_level AS (
    SELECT
        "State",
        ROUND(AVG(avg_city_rating)::NUMERIC, 2) AS avg_state_rating
    FROM
        city_level
    GROUP BY
        "State"
)
SELECT
    *
FROM
    state_level
ORDER BY
    avg_state_rating DESC
LIMIT 10;
"""

df_q25 = pd.read_sql_query(query_25, con=engine)
df_q25

State  avg_state_rating
0    SD              4.50
1    ND              4.33
2    CO              4.27
3    WI              4.22
4    MA              4.17
5    UT              4.00
6    VA              4.00
7    KS              3.84
8    NC              3.80
9    SC              3.76

### 📊 Query 25 Summary

The states with the highest average hospital ratings, considering only cities with at least three hospitals, include **South Dakota (4.50)**, **North Dakota (4.33)**, and **Colorado (4.27)**. This suggests strong healthcare quality in these states within urban areas having sufficient hospital density. This layered analysis helps identify regions with consistently high-performing hospitals, filtering out smaller or less representative cities.

---
### 🧮 Query 26: Compare Hospitals Across States Using UNION and EXCEPT

**Business Question**  
How can we compare hospital lists between Illinois and Texas using set operations?

**Why this matters**  
Set operations like `UNION` and `EXCEPT` enable analysts to merge or differentiate data groups efficiently. Here, we combine hospitals from both states and also identify those unique to Illinois.

**SQL Concepts Introduced**  
- Use of `UNION` to combine datasets without duplicates  
- Use of `EXCEPT` to find differences between datasets  
- Ordering and limiting results for concise output

In [29]:
# Query 26a: Combine hospitals from Illinois and Texas using UNION
query_26_union = """
SELECT 
    "Facility ID",
    "Facility Name",
    "City/Town",
    "State"
FROM 
    hospital_info
WHERE 
    "State" = 'IL'

UNION

SELECT 
    "Facility ID",
    "Facility Name",
    "City/Town",
    "State"
FROM 
    hospital_info
WHERE 
    "State" = 'TX'
ORDER BY 
    "Facility Name"
LIMIT 10;
"""

df_q26_union = pd.read_sql_query(query_26_union, con=engine)

# Query 26b: Find hospitals in Illinois but not in Texas using EXCEPT
query_26_except = """
SELECT 
    "Facility Name"
FROM 
    hospital_info
WHERE 
    "State" = 'IL'

EXCEPT

SELECT 
    "Facility Name"
FROM 
    hospital_info
WHERE 
    "State" = 'TX'
ORDER BY 
    "Facility Name"
LIMIT 10;
"""

df_q26_except = pd.read_sql_query(query_26_except, con=engine)

# Display side-by-side using HTML

display(HTML(
    '<div style="display:flex; gap:40px;">'
    f'<div style="flex:1;"><h4>UNION (IL + TX)</h4>{df_q26_union.to_html(index=False)}</div>'
    f'<div style="flex:1;"><h4>EXCEPT (IL - TX)</h4>{df_q26_except.to_html(index=False)}</div>'
    '</div>'
))

### 📊 Query 26 Summary

The `UNION` query combines hospitals from **Illinois** and **Texas**, showing a merged list of facilities from both states sorted by name. The `EXCEPT` query identifies hospitals present in **Illinois** but **not in Texas**, highlighting state-specific healthcare providers. These set operations are powerful for comparing and contrasting data subsets efficiently.

---
### 🧮 Query 27: Compare Hospital Ratings Within States Using LAG()

**Business Question**  
How does a hospital's current overall rating compare to the previous hospital's rating alphabetically within the same state?

**Why this matters**  
This query demonstrates how to use window functions to compare adjacent rows, which is useful for detecting trends or changes within groups, such as states in this dataset.

**SQL Concepts Introduced**  
- Window function `LAG()` for accessing prior row values within partitions  
- `PARTITION BY` and `ORDER BY` clauses within window functions  
- Casting and ordering results

In [30]:
# Query 27: Use LAG() window function to compare current hospital rating with previous one in the same state
query_27 = """
SELECT 
    "Facility ID",
    "Facility Name",
    "State",
    "Hospital overall rating"::FLOAT AS current_rating,
    LAG("Hospital overall rating"::FLOAT) OVER (
        PARTITION BY "State"
        ORDER BY "Facility Name"
    ) AS previous_rating
FROM 
    hospital_info
WHERE 
    "Hospital overall rating" IS NOT NULL
ORDER BY 
    "State", 
    "Facility Name"
LIMIT 20;
"""

df_q27 = pd.read_sql_query(query_27, con=engine)
df_q27

Facility ID                             Facility Name State  \
0       020026              ALASKA NATIVE MEDICAL CENTER    AK   
1       020017                  ALASKA REGIONAL HOSPITAL    AK   
2       020008                BARTLETT REGIONAL HOSPITAL    AK   
3       020024        CENTRAL PENINSULA GENERAL HOSPITAL    AK   
4       020012               FAIRBANKS MEMORIAL HOSPITAL    AK   
5       020006            MAT-SU REGIONAL MEDICAL CENTER    AK   
6       020001          PROVIDENCE ALASKA MEDICAL CENTER    AK   
7       010036                          ANDALUSIA HEALTH    AL   
8       010079                 ATHENS LIMESTONE HOSPITAL    AL   
9       010083                            BALDWIN HEALTH    AL   
10      010149               BAPTIST MEDICAL CENTER EAST    AL   
11      010023              BAPTIST MEDICAL CENTER SOUTH    AL   
12      01014F              BIRMINGHAM VA MEDICAL CENTER    AL   
13      010139          BROOKWOOD BAPTIST MEDICAL CENTER    AL   
14      010164               COOSA VALLEY MEDICAL CENTER    AL   
15      010131                  CRESTWOOD MEDICAL CENTER    AL   
16      010035           CULLMAN REGIONAL MEDICAL CENTER    AL   
17      010021                       DALE MEDICAL CENTER    AL   
18      010092               DCH REGIONAL MEDICAL CENTER    AL   
19      010085  DECATUR MORGAN HOSPITAL - DECATUR CAMPUS    AL   

    current_rating  previous_rating  
0              3.0              NaN  
1              2.0              3.0  
2              4.0              2.0  
3              1.0              4.0  
4              2.0              1.0  
5              3.0              2.0  
6              3.0              3.0  
7              3.0              NaN  
8              2.0              3.0  
9              5.0              2.0  
10             3.0              5.0  
11             2.0              3.0  
12             3.0              2.0  
13             2.0              3.0  
14             4.0              2.0  
15             3.0              4.0  
16             3.0              3.0  
17             2.0              3.0  
18             1.0              2.0  
19             3.0              1.0

### 📊 Query 27 Summary  
This query uses the `LAG()` window function to compare each hospital's current overall rating with the previous hospital’s rating within the same state, ordered alphabetically by hospital name. This helps identify rating changes or inconsistencies between neighboring facilities in each state. For example, in Alaska (AK), *ALASKA REGIONAL HOSPITAL* has a current rating of **2.0** compared to the previous hospital’s **3.0**.

---
### 🧮 Query 28: Hospital Ratings Distribution by Ownership Type

**Business Question**  
How are hospital ratings distributed across different hospital ownership types?

**Why this matters**  
Understanding the spread of hospital quality ratings by ownership category can highlight patterns of performance and help stakeholders identify areas needing attention or investment.

**SQL Concepts Introduced**  
- Pivot-style aggregation using `SUM` with `CASE WHEN` for counting conditional categories
- Grouping and ordering results for summary statistics

In [31]:
# Query 28: Pivot-style aggregation of hospital ratings by ownership type
query_28 = """
SELECT
    "Hospital Ownership",
    COUNT(*) AS total_hospitals,
    SUM(CASE WHEN "Hospital overall rating" = '1' THEN 1 ELSE 0 END) AS one_star,
    SUM(CASE WHEN "Hospital overall rating" = '2' THEN 1 ELSE 0 END) AS two_star,
    SUM(CASE WHEN "Hospital overall rating" = '3' THEN 1 ELSE 0 END) AS three_star,
    SUM(CASE WHEN "Hospital overall rating" = '4' THEN 1 ELSE 0 END) AS four_star,
    SUM(CASE WHEN "Hospital overall rating" = '5' THEN 1 ELSE 0 END) AS five_star
FROM
    hospital_info
GROUP BY
    "Hospital Ownership"
ORDER BY
    total_hospitals DESC;
"""

df_q28 = pd.read_sql_query(query_28, con=engine)
df_q28

Hospital Ownership  total_hospitals  one_star  \
0                Voluntary non-profit - Private             2280       102   
1                                   Proprietary             1060        87   
2   Government - Hospital District or Authority              512        19   
3                            Government - Local              396        21   
4                  Voluntary non-profit - Other              361        20   
5                 Voluntary non-profit - Church              272        10   
6                            Government - State              208         5   
7                Veterans Health Administration              132        10   
8                                     Physician               74         1   
9                          Government - Federal               42         0   
10                        Department of Defense               32         0   
11                                       Tribal               15         0   

    two_star  three_star  four_star  five_star  
0        247         432        420        195  
1        132         133         83         41  
2         55          48         38         12  
3         38          35         34         13  
4         39          66         74         42  
5         39          59         75         28  
6          9          13          4          5  
7         14          23         27         39  
8          6           4          1          5  
9          4           2          2          0  
10         0           0          0          0  
11         0           1          0          0

### 📊 Query 28 Summary  
This query pivots hospital overall ratings into columns by ownership type, showing the count of hospitals for each star rating (1 to 5 stars). For example, **Voluntary non-profit - Private** owns the largest number of hospitals (**2280**), with a broad distribution of ratings across the scale. Other ownership types such as **Proprietary** and **Government - Hospital District or Authority** also show distinct rating distributions, highlighting differences in hospital quality by ownership.

---
## 🎯 Conclusion

This notebook demonstrates a comprehensive and practical application of SQL skills on a real-world hospital dataset. Through 28 thoughtfully designed queries, I have explored a variety of important aspects such as hospital distribution, rating patterns, ownership impact, and performance tiers across states and cities.

Key takeaways include:
- Geographic and ownership-based variation in hospital counts and quality ratings.
- Identification of data gaps like missing ratings by state.
- Advanced SQL techniques such as window functions (`RANK()`, `LAG()`, `NTILE()`), Common Table Expressions (CTEs), set operations (`UNION`, `EXCEPT`), and pivot-style aggregations.
- Realistic business questions answered through SQL, showcasing my ability to extract actionable insights for stakeholders.

This end-to-end exploration highlights my proficiency with SQL fundamentals, advanced querying, and data storytelling—key skills for any data analyst or data scientist role. The structured approach also prepares me to build dashboards and reports that communicate findings effectively to business stakeholders.

I look forward to applying these skills in real-world projects and continuing to deepen my expertise with more complex datasets and use cases.